src/cli/init.js

```diff
@@ -0,0 +1,183 @@
+import {specToProject} from "../plugins/github/specToProject";
+import * as NullUtil from "../util/null";
+
+function usage(print: (string) => void): void {
+  print(
+    dedent`\
+    usage: sourcecred init [--github GITHUB_SPEC [...]]
+                           [--discourse DISCOURSE_URL]
+                           [--force]
+                           [--print]
+           sourcecred init --help
+
+    Sets up a new SourceCred instance, by creating a SourceCred project
+    configuration, and saving it to 'sourcecred.json' within the current
+    directory.
+
+    Zero or more GitHub specs may be provided; each GitHub spec can be of the
+    form OWNER/NAME (as in 'torvalds/linux') for loading a single repository,
+    or @owner (as in '@torvalds') for loading all repositories owned by a given
+    account. If any GitHub specs are present, then the SOURCECRED_GITHUB_TOKEN
+    environment variable is required.
+
+    A discourse url may be provided. The discourse url must be the full url of
+    a valid Discourse server, as in 'https://discourse.sourcecred.io'.
+
+    All of the GitHub specs, and the Discourse specification (if it exists)
+    will be combined into a single project, which is written to
+    sourcecred.json. The file may be manually modified to activate other
+    advanced features, such as identity map resolution.
+
+    Arguments:
+        --github GITHUB_SPEC
+            A specification (in form 'OWNER/NAME' or '@OWNER') of GitHub
+            repositories to load.
+
+        --discourse DISCOURSE_URL
+            The url of a Discourse server to load.
+
+        --force
+            If provided, sourcecred init will overwrite pre-existing
+            sourcecred.json files. Otherwise, the command will refuse to
+            overwite pre-existing files and fail.
+
+        --print
+            If provided, sourcecred init will print the project to stdout
+            rather than writing it to sourcecred.json. Mostly used for testing
+            purposes, as a SourceCred instance is only valid if the file is
+            saved as sourcecred.json. If this flag is set, it supercedes
+            --force.
+
+        --help
+            Show this help message and exit, as 'sourcecred help init'.
+
+    Environment variables:
+        SOURCECRED_GITHUB_TOKEN
+            API token for GitHub. This should be a 40-character hex
+            string. Required if you provide GitHub specs.
+
+            To generate a token, create a "Personal access token" at
+            <https://github.com/settings/tokens>. When loading data for
+            public repositories, no special permissions are required.
+            For private repositories, the 'repo' scope is required.
+    `.trimRight()
+  );
+}
+
+function die(std, message) {
+  std.err("fatal: " + message);
+  std.err("fatal: run 'sourcecred help init' for help");
+  return 1;
+}
+
+const initCommand: Command = async (args, std) => {
```

Comment:
Like with previous PR, suggest separating the:
- cli command handling
- `Project` generating logic
- I/O or printing of a `Project`

Reply:
So, I tried this, and the resultant code felt much more complicated, difficult to reason about and maintain. As an example, I tried to factor the argument parsing logic out of the main IO flow of the project, but then found myself dealing with issues like that the command should print usage and exit as soon as it finds the `--help` flag, rather than potentially error-ing due to malformed flags.

As I went down this (and other rabbit holes), I felt that I was solving a puzzle rather than a problem (to borrow a quote from [a Hickey talk you linked](https://www.youtube.com/watch?v=2V1FtfBDsLU)). IMO, having a complected 100 line CLI script written in an imperative mood is not a problem.

Here is what I came up with when I tried to pull it apart. Let me know if you would have preferred this version.

```js
const initCommand: Command = async (args, std) => {
  const maybeParsedArgs = parseArguments(args);
  if (maybeParsedArguments.type === "FAILURE") {
    return die(maybeParsedArgs.failure);
  }
  const parsedArgs = maybeParsedArgs.result;
  if (parsedArgs.wantsHelp) {
    usage(std.out);
    return 0;
  }

  const maybeProject = await generateProject(parsedArgs, Common.githubToken());
  if (maybeProject.type === "FAILURE") {
    return die(maybeProject.failure);
  }
  const project = maybeProject.result;

  const dir = process.cwd();
  const projectFilePath = path.join(dir, "sourcecred.json");
  if ((await fs.exists(projectFilePath)) && !withForce) {
    return die(std, `refusing to overwrite sourcecred.json without --force`);
  }
  const basename = path.basename(dir);

  const githubToken = Common.githubToken();
  if (githubToken == null && githubSpecs.length > 0) {
    return die(
      std,
      "tried to load GitHub specs, but no GitHub token provided."
    );
  }
  let repoIds: RepoId[] = [];
  for (const spec of githubSpecs) {
    const subproject = await specToProject(spec, NullUtil.get(githubToken));
    repoIds = [...repoIds, ...subproject.repoIds];
  }

  const discourseServer: DiscourseServer | null = discourseUrl
    ? {serverUrl: discourseUrl}
    : null;
  const project: Project = {
    id: basename,
    repoIds,
    discourseServer,
    identities: [],
  };

  const projectJson = projectToJSON(project);
  await fs.writeFile(projectFilePath, JSON.stringify(projectJson, null, 2));

  std.out("Wrote project file to 'sourcecred.json'");

  return 0;
};
```

You can find the [implementation here](https://github.com/sourcecred/sourcecred/blob/monadic-init/src/cli/init.js). (Sidenote: how do you push a single commit to GitHub without making a vestigial branch for it?)

Reply:
> (Sidenote: how do you push a single commit to GitHub without making a vestigial branch for it?)

Afaik, you can't. Typically I link to commits after using a branch to push them.

Reply:
I've added an example of what I have in mind: https://github.com/sourcecred/sourcecred/commit/9830bb4e896d2ba34e7ab5a1898880b49aef5250

Notice how this makes the functions `generateProject` and `outputProject` correct even outside the CLI context. And how much less complex `initCommand` is.

Reply:
> Sidenote: how do you push a single commit to GitHub without making a
> vestigial branch for it?

Use `git push origin HEAD:refs/heads/new-branch-name`, or replace `HEAD`
with any ref.

This obviously creates a branch on GitHub’s side, but I presume that
you’re asking how to avoid creating a local branch. This is how.

I use this dozens of times per day. I don’t use branches for individual
pull requests (instead, I use them for actual *branches* of sequences of
changes when this is a helpful thing to do), so this is my standard
workflow.

Reply:
Didn't come here for this, but it made my day. Thanks @wchargin